### PR TITLE
Questionnaire different ≤ 4.2 and ≥ 4.3 activity completion behat tests …

### LINKED ISSUE #522: Have the MOODLE_401_STABLE branch pass automated tests
Currently, the MOODLE_401_STABLE branch does not pass automated tests because master doesn't play along, since activity completion is set differently, now.

### PR DESCRIPTION
**Pull request by lucaboesch**
… fix #522 .

Dear @mchurchward this pull request lets pass the automated tests again.

Best,
Luca

### CODE REVIEW

**Comment by rezeau** [comment]
Just tested this PR on my local Moodle site running Moodle 4.3. PR Works as expected, the first scenario (_Check questionnaire completion feature in web for Moodle ≤ 4.2_) is skipped and the "_Check questionnaire completion feature in web for Moodle ≥ 4.3_" scenario is passed OK.
A _very minor_ grievance: I am running those behat tests on my Windows box, with profile=geckodriver and the test begins to show in an Acceptance  text site window but, when scenario 1 is skipped, then that window closes and does not re-open. However, if I add the --format=pretty parameter I can follow all the steps in my Windows shell window, so that's OK.

**Comment by mchurchward** [comment]
@lucaboesch thanks for this. Do you know why we need to add "xpath" for 4.3? This seems to be something that would not be desirable.

**Comment by lucaboesch** [comment]
And I click on "3" radio wouldn't execute any more, @mchurchward, I tried it...

**Comment by mchurchward** [comment]
But it does work in other places, like:

- add_feedback.feature
- add_multi_feedback_with_sections.feature

That makes me think there is something wrong with the display of that radio. There may need to be a fix in the display code.

**Comment by mchurchward** [comment]
Also, the error that was [highlighted here](https://github.com/PoetOS/moodle-mod_questionnaire/actions/runs/6436573851/job/17480171307?pr=520#step:17:221) was for "Completion tracking". So the radio error came up after that was fixed?

**Comment by mchurchward** [comment]
@lucaboesch I just tested this locally, and the only change it needs is for the "Completion tracking". The radio buttons work fine.

**Comment by lucaboesch** [comment]
It would be of great help if you could paste in the modified lines here, @mchurchward

**Comment by lucaboesch** [comment]
I had it run again.
The Moodle Plugin CI run error states

```
001 Scenario: Check questionnaire completion feature in web for Moodle ≥ 4.3. # /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/mod/questionnaire/tests/behat/questionnaire_activity_completion.feature:70
      And I click on "Yes" "radio"                                            # /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/mod/questionnaire/tests/behat/questionnaire_activity_completion.feature:111
        Behat\Mink\Exception\UnsupportedDriverActionException: Behat\Mink\Driver\BrowserKitDriver supports clicking on links and submit or reset buttons only. But "input" provided in /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/vendor/behat/mink-browserkit-driver/src/BrowserKitDriver.php:468
```

File: tests/behat/questionnaire_activity_completion.feature [hunk 1]
```diff
@@ -65,3 +66,51 @@ Feature: View activity completion information in the questionnaire activity
...
+      | Possible answers | 1=One,2=Two,3=Three,4=Four |
+    Then I should see "[Radio Buttons] (Q2)"
+    And I add a "Text Box" question and I fill the form with:
+      | Question Name    | Q8              |
+      | No               | n               |
+      | Input box length | 10              |
+      | Max. text length | 15              |
+      | Question Text    | Enter some text |
+    Then I should see "[Text Box] (Q8)"
+    And I am on the "Test questionnaire completion" "questionnaire activity editing" page
```
Comment (mchurchward) [inline tests/behat/questionnaire_activity_completion.feature:73]: I think the lines below can be simplified to:

    And I expand all fieldsets
    And I set the following fields to these values:
      | Add requirements                                      | 1 |
      | Student must submit this questionnaire to complete it | 1 |

This is from examples in core.

File: tests/behat/questionnaire_activity_completion.feature [hunk 2]
```diff
@@ -65,3 +66,51 @@ Feature: View activity completion information in the questionnaire activity
...
+
+    And I am on the "Test questionnaire completion" "questionnaire activity" page
+    Then I should see "You are not eligible to take this questionnaire."
+
+    And I am on the "Test questionnaire completion" "questionnaire activity" page logged in as "student1"
+    And I click on "Answer the questions..." "link"
+    Then I should see "Are you still in School?"
+    And I should see "Select one choice"
+    And I should see "Enter some text"
+    And I set the field with xpath "//input[@type='radio' and @id='auto-rb0001']" to "1"
```
Comment (mchurchward) [inline tests/behat/questionnaire_activity_completion.feature]: The code:
    And I click on "Yes" "radio"
    And I click on "Three" "radio"
 still works for me running Behat locally. I don't think this and the next change are needed.

**Review by mchurchward — CHANGES_REQUESTED** [review]

File: tests/behat/questionnaire_activity_completion.feature (hunk 2 above)
Comment (mchurchward) [reply tests/behat/questionnaire_activity_completion.feature]: I wonder if its due to "master". The CI should be changed to MOODLE_403_STABLE now that it has been released.

**Comment by mchurchward** [comment]
> I had it run again. The Moodle Plugin CI run error states
> 
> ```
> 001 Scenario: Check questionnaire completion feature in web for Moodle ≥ 4.3. # /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/mod/questionnaire/tests/behat/questionnaire_activity_completion.feature:70
>       And I click on "Yes" "radio"                                            # /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/mod/questionnaire/tests/behat/questionnaire_activity_completion.feature:111
>         Behat\Mink\Exception\UnsupportedDriverActionException: Behat\Mink\Driver\BrowserKitDriver supports clicking on links and submit or reset buttons only. But "input" provided in /home/runner/work/moodle-mod_questionnaire/moodle-mod_questionnaire/moodle/vendor/behat/mink-browserkit-driver/src/BrowserKitDriver.php:468
> ```

I wonder if that is a Chrome issue with Behat? Maybe try changing the CI behat line:
        run: moodle-plugin-ci behat
(drop the --profile chrome part)?

**Comment by mchurchward** [comment]
@lucaboesch I figured out the problem. Radio buttons should not use "I click". They need a "Set field to" instead. But, they do work if "@javascript" is specified, and it wasn't for the new 4.3 scenario. In any case, the correct fix is to use "set field to".
See my [commit here](https://github.com/PoetOS/moodle-mod_questionnaire/commit/c43b770932e68fb167318a92510cf79398acf6c7). It does the "set field" and additionally I simplified the scenarios by moving the duplicate steps to the background. If you want to update your PR accordingly, I will merge it in.

**Comment by lucaboesch** [comment]
Yeah, @mchurchward, but the point was to make the test passing for Moodle ≥ 4.3 as well as Moodle ≤ 4.2 so you don't resolve in leaving out those versions in ci.yml.
This updated pull request fits all needs and supports PHP 7.4, 8.0 and 8.1.

Best,
Luca

**Review by mchurchward — APPROVED** [review]

**Comment by mchurchward** [comment]
Thanks Luca. Of course once we add PHP8.2 to the CI, there will be a huge mess. :-0

**Pull request merged**